### ISSUE TITLE
optional handling in object

### ISSUE THREAD

**Issue by jordan-boyer** [thoughtspile/banditypes#3]
given the following snippet 

```ts
const person = object({
    name: string().or(optional()),
    age: number(),
  });

type Person = Infer<typeof person>;
```
we have the resulting person type

```ts
type Person = {
    name: string | undefined;
    age: number;
}
```

where we could have 

```ts
type Person = {
    name?: string | undefined;
    age: number;
}
```

what do you think ?

**Comment by adversinc** [comment]
I second that. Infer<> does not generate the optional parameters so the following code results an error:

```typescript
const person = object({
    name: string().or(optional()),
    age: number().or(optional()),
  });

type Person = Infer<typeof person>;

function updatePerson(personData: Person = {}) { ... }
```

It would be great to get it fixed.

**Comment by thoughtspile** [comment]
Ouch, looks like superstruct typetest suite has failed me: https://github.com/thoughtspile/banditypes/blob/0b992f7e9ea04894fde8c1215eb6e85e1fcaefea/tests/typings/optional.ts#L6-L12

The inferred type is assignable to the target one, but they're not equal.

**Comment by jordan-boyer** [comment]
If you want a better test you can add this `"exactOptionalPropertyTypes": true` and write the test like this:
```ts
test<{
  a?: number
  b: string
}>(object({
  a: number().or(optional()),
  b: string(),
}))
```

which typescript complain without a proper fix for the optional on object

**Comment by thoughtspile** [comment]
Fixed in `0.2.3`

I also [updated the typetest suite](https://github.com/thoughtspile/banditypes/pull/6) for strict checks, and opened an [issue in superstruct](https://github.com/ianstormtaylor/superstruct/issues/1174) in case they'd like to avoid problems like these.
